### PR TITLE
Add verification of replication tasks in force replication

### PR DESCRIPTION
**Pull request by hehaifengcn**
**What changed?**
Add a verification step to check if generated workflow executions exist on target cluster. 

**Why?**
To ensure all generated replication tasks have been successfully applied on target cluster.

**How did you test it?**
Unit tests + cluster tests

**Potential risks**

**Is hotfix candidate?**
No

### CODE REVIEW

File: service/worker/migration/activities.go [hunk 1]
```diff
@@ -482,3 +540,194 @@ func (a *activities) SeedReplicationQueueWithUserDataEntries(ctx context.Context
...
+	for i := 0; i < len(request.Executions); i++ {
+		r := &detail.Results[i]
+		if r.isCompleted() {
+			continue
+		}
+
+		we := request.Executions[i]
+		tags := []tag.Tag{tag.WorkflowType(forceReplicationWorkflowName), tag.WorkflowNamespaceID(request.NamespaceID), tag.WorkflowID(we.WorkflowId), tag.WorkflowRunID(we.RunId)}
+
+		resp, err := a.historyClient.DescribeMutableState(ctx, &historyservice.DescribeMutableStateRequest{
```
Comment (wxing1292) [inline service/worker/migration/activities.go:556]: `a.historyClient` -> `a.localHistoryClient`?

Comment (wxing1292) [reply service/worker/migration/activities.go:556]: or source history client

File: service/worker/migration/activities.go [hunk 2]
```diff
@@ -482,3 +540,194 @@ func (a *activities) SeedReplicationQueueWithUserDataEntries(ctx context.Context
...
+			NamespaceId: request.NamespaceID,
+			Execution:   &we,
+		})
+
+		switch err.(type) {
+		case nil:
+			if resp.GetCacheMutableState().GetExecutionState().GetState() == enumsspb.WORKFLOW_EXECUTION_STATE_ZOMBIE {
```
Comment (wxing1292) [inline service/worker/migration/activities.go:563]: `if resp.GetCacheMutableState()` this may return nil, use `GetDatabaseMutableState`

**Review by wxing1292 — APPROVED** [review]
overall idea LGTM

implementation can be improved, by e.g. break giant for loop into for loop and function invocations

File: service/worker/migration/activities.go (hunk 1 above)
Comment (hehaifengcn) [reply service/worker/migration/activities.go:556]: i think `historyClient` implicitly mean `local` as other where in the codebase. I can change it if you feel strong about it.

File: service/worker/migration/activities.go (hunk 2 above)
Comment (hehaifengcn) [reply service/worker/migration/activities.go:563]: what are the difference between 2? `GetDatabaseMutableState` also just read from the object?

Comment (wxing1292) [reply service/worker/migration/activities.go]: GetCacheMutableState contains the cached version of mutable state, which can be nil (not cached)
GetDatabaseMutableState directly load mutable state from DB

File: service/worker/migration/activities.go [hunk 3]
```diff
@@ -482,3 +541,197 @@ func (a *activities) SeedReplicationQueueWithUserDataEntries(ctx context.Context
...
+			Execution:   &we,
+		})
+
+		switch err.(type) {
+		case nil:
+			if resp.GetDatabaseMutableState().GetExecutionState().GetState() == enumsspb.WORKFLOW_EXECUTION_STATE_ZOMBIE {
+				a.forceReplicationMetricsHandler.Counter(metrics.EncounterZombieWorkflowCount.GetMetricName()).Record(1)
+				a.logger.Info("createReplicationTasks skip Zombie workflow", tags...)
+
+				r.Status = VERIFY_SKIPPED
```
Comment (yux0) [inline service/worker/migration/activities.go:568]: Seems like it is not just skip verify, it also skips generate replication task. In addition, should we filter this in the force replication API?

File: service/worker/migration/activities.go [hunk 4]
```diff
@@ -51,6 +53,65 @@ import (
...
+	VerifyStatus int
+	VerifyResult struct {
+		Status VerifyStatus
+		Reason string
+	}
+
+	replicationTasksHeartbeatDetails struct {
+		Results                       []VerifyResult
+		CheckPoint                    time.Time
+		LastNotFoundWorkflowExecution commonpb.WorkflowExecution
```
Comment (yux0) [inline service/worker/migration/activities.go:66]: What is the purpose of this? Should we use a fixed length slice?

File: service/worker/migration/activities.go [hunk 5]
```diff
@@ -482,3 +541,197 @@ func (a *activities) SeedReplicationQueueWithUserDataEntries(ctx context.Context
...
+		details.Results = make([]VerifyResult, len(request.Executions))
+		details.CheckPoint = time.Now()
+		activity.RecordHeartbeat(ctx, details)
+	}
+
+	start := time.Now()
+	if err := a.createReplicationTasks(ctx, request, &details); err != nil {
+		return err
+	}
+	a.forceReplicationMetricsHandler.Timer(metrics.CreateReplicationTasksLatency.GetMetricName()).Record(time.Since(start))
```
Comment (yux0) [inline service/worker/migration/activities.go:680]: emit the latency inside createReplicationTasks()?

File: service/worker/migration/activities.go [hunk 6]
```diff
@@ -482,3 +541,197 @@ func (a *activities) SeedReplicationQueueWithUserDataEntries(ctx context.Context
 		activity.RecordHeartbeat(ctx, heartbeatDetails)
 	}
 }
+
+func (a *activities) createReplicationTasks(ctx context.Context, request *genearteAndVerifyReplicationTasksRequest, detail *replicationTasksHeartbeatDetails) error {
+	rateLimiter := quotas.NewRateLimiter(request.RPS, int(math.Ceil(request.RPS)))
+
+	for i := 0; i < len(request.Executions); i++ {
+		r := &detail.Results[i]
+		if r.isCompleted() {
```
Comment (yux0) [inline service/worker/migration/activities.go:550]: Is this a valid case? Do you plan to call createReplicationTasks() then verify() and then createReplicationTasks()?

File: common/metrics/metric_defs.go [hunk 7]
```diff
@@ -1563,6 +1563,12 @@ var (
 	ScheduleCancelWorkflowErrors                      = NewCounterDef("schedule_cancel_workflow_errors")
 	ScheduleTerminateWorkflowErrors                   = NewCounterDef("schedule_terminate_workflow_errors")
 
+	// Force replication
```
Comment (yux0) [inline common/metrics/metric_defs.go:1566]: Do we care about verification failure?

File: service/worker/migration/activities.go (hunk 3 above)
Comment (hehaifengcn) [reply service/worker/migration/activities.go:568]: yes. are u suggesting that we should not skips generate replication task for Zombie WF? 
what does "should we filter this in the force replication API?" mean?

File: service/worker/migration/activities.go (hunk 4 above)
Comment (hehaifengcn) [reply service/worker/migration/activities.go:66]: VerifyResult keeps the status of replication tasks for each execution task from input, which is a variable-length array:

File: common/metrics/metric_defs.go (hunk 7 above)
Comment (hehaifengcn) [reply common/metrics/metric_defs.go:1566]: i think we can use activity failure metrics. we can always add later if needed.

File: service/worker/migration/activities.go (hunk 3 above)
Comment (yux0) [reply service/worker/migration/activities.go:573]: synced offline. Not do this at current stage.

**Review by yux0 — APPROVED** [review]

**Pull request merged**